RHSTOR-1554: Migrate object service dashboard tests

rexagod opened the pull request:
Based-off-of: https://github.com/openshift/console/pull/7815

Fixes: https://issues.redhat.com/browse/RHSTOR-1554

rexagod commented:
@bipuladh I've made the changes suggested in the review.

PTAL. Thanks!

rexagod commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,149 @@
+  beforeEach(() => {
+    cy.visit('/');
+    cy.byLegacyTestID('horizontal-link-Object Service').click();
+    cy.wait(10 * SECOND);
```
Confirmed this a couple of times and both `it` need a 10 second buffer time for the page to load.

bipuladh replied:
Wait for the elements that you want to test. For example if you want to test the details card then wait on the details card to load. Don't wait arbitrarily.

rexagod replied:
Right, that's better. Sending a patch soon.

rexagod commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,149 @@
+  beforeEach(() => {
+    cy.visit('/');
+    cy.byLegacyTestID('horizontal-link-Object Service').click();
+    cy.wait(10 * SECOND);
```
> Confirmed this a couple of times and both it need a 10 second buffer time for the page to load.

Same here.

rexagod commented:
Tests look good on local.

![image](https://user-images.githubusercontent.com/33557095/109624443-3a4cf480-7b64-11eb-984a-25ce970f5701.png)

rexagod commented:
/retest

rexagod commented:
/retest

vbnrh commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,149 @@
+  });
+
+  it('Create an Object Bucket Claim and test equality', () => {
+    cy.wait(10 * SECOND)
```
As per best practices, arbitrary wait times are not allowed. This is one of the issues i faced in my PR, you probably need to look into this

rexagod replied:
Thanks for pointing this out! Sending a patch soon based on [this](https://docs.cypress.io/guides/references/best-practices.html#Unnecessary-wait-for-cy-get) soon.

rexagod replied:
It seems the above method is only for routes. Using `{ timeout: __ }` instead.

rexagod commented:
I may very well be wrong but it seems like the `/dashboards/object-service` page loads twice, which is the reason why `cy.wait` is necessary. Not sure why though.

rexagod commented:
@bipuladh @vbnrh Any leads on how to fetch the current state for these three from the CLI? Would be a great help! Thanks!

![image](https://user-images.githubusercontent.com/33557095/109763104-96248580-7c17-11eb-8a68-eafe7c79c1ec.png)

rexagod commented:
Okay I think I got it. 

![image](https://user-images.githubusercontent.com/33557095/109763804-a2f5a900-7c18-11eb-9fed-54109aee13ad.png)

rexagod commented:
Tried this with @bipuladh's CLI suggestion, worked on the first try! :)

![image](https://user-images.githubusercontent.com/33557095/109765458-e4875380-7c1a-11eb-9dbf-589fee128623.png)

rexagod commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  });
+
+  it('Test at least one noobaa bucket is present', () => {
+    cy.byTestID('1-obc-test')
```
This assumes exactly `1 NooBaa Bucket` for the test, only for the time being. Will replace this with a better alternative once I figure out how to get the number of "NooBaa Bucket" via CLI.

bipuladh replied:
Noobaa buckets are queried via Prometheus.

rexagod replied:
Fixed in https://github.com/openshift/console/commit/41a827f93d807aaf95c05dba3cf89595f6838fa6.

bipuladh commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  cy.install();
+};
+
+describe('Tests Buckets Card', () => {
```
You don't need multiple describe blocks. Just use one describe block and separate different cards into various 'it' blocks.

rexagod replied:
Do you mean that I should covert existing additional describe blocks into it blocks and remove the sub-it blocks that are inside those? I've pushed a patch that simply removes the additional describes but keeps their it blocks.

Let me know what changes you have in mind and I'll push them.

bipuladh commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  });
+
+  it('Create an Object Bucket Claim and test equality', () => {
+    cy.exec(`kubectl get ObjectBucketClaims -A | wc -l`).then((outObj) => {
```
```suggestion
    cy.exec(`kubectl get ObjectBucketClaims -A | wc -l`).then(({stdout})) => {
```

bipuladh reviewed (changes requested):
Need to look in-depth, for now we can start with these changes

rexagod commented:
/retest

rexagod commented:
Fixed the hardcoded "x NooBaa Bucket" in this commit (https://github.com/openshift/console/pull/8270/commits/41a827f93d807aaf95c05dba3cf89595f6838fa6).

rexagod commented:
- https://github.com/openshift/console/pull/8270/commits/e15769c8d28d1c5a5cd032c0678b4b2198af9e83: Moved describe blocks to it blocks, and it blocks to `cy.log`.
- https://github.com/openshift/console/pull/8270/commits/b7710aeefe6958ee43bcc0dc7f1d44febc753f8d: Moved some constants to `consts.ts`.

rexagod commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,98 @@
+
+describe('Tests Buckets, Status, Object Storage Efficiency, and Resource Providers Cards', () => {
+  before(() => {
+    installStorageCluster();
```
Move `installStorageCluster` in here.

rexagod commented on frontend/packages/console-shared/src/components/dashboard/inventory-card/InventoryItem.tsx:
```diff
@@ -118,7 +118,15 @@ export const InventoryItem: React.FC<InventoryItemProps> = React.memo(
-          {TitleComponent ? <TitleComponent>{titleMessage}</TitleComponent> : titleMessage}
+          {TitleComponent ? (
+            <TitleComponent>
+              <div data-test={`${parseInt(titleMessage, 10).toString()}-obc-test`}>
```
Remove.

rexagod commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,98 @@
+    cy.byLegacyTestID('horizontal-link-Object Service').click();
+  });
+
+  it('Tests Buckets Cards', () => {
```
TODO.

rexagod commented:
/test e2e-gcp-console

rexagod commented:
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/8270/pull-ci-openshift-console-master-e2e-gcp-console/1372817923416002560/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/cypress/screenshots/

rexagod commented:
/test e2e-gcp-console

bipuladh reviewed (approved):
/lgtm
/assign @rawagner

rexagod commented on frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:
```diff
@@ -0,0 +1,98 @@
+
+describe('Tests Buckets, Status, Object Storage Efficiency, and Resource Providers Cards', () => {
+  before(() => {
+    installStorageCluster();
```
I think you're viewing an older commit. I've done this in the [latest commit](https://github.com/openshift/console/pull/8270/files#diff-f5ff9081f8c3177bae430651bc5e1879b0a31216c50bbae818e41d100e46982dR5).

rexagod commented:
@bipuladh Can I go ahead and squash these commits now?

rexagod commented:
/test e2e-gcp-console

rawagner commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -5,28 +5,28 @@ import { DashboardCardPopupLink } from '../dashboard-card/DashboardCardLink';
 
 const HealthItem: React.FC<HealthItemProps> = React.memo(
-  ({ className, state, title, details, popupTitle, noIcon = false, children }) => {
+  ({ className, state, title, details, popupTitle, noIcon = false, dataTest, children }) => {
```
I dont see any code calling `HealthItem` with this `dataTest` prop .. ?

rexagod replied:
Right, fixed!

rawagner commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -5,28 +5,28 @@ import { DashboardCardPopupLink } from '../dashboard-card/DashboardCardLink';
 
     return (
-      <div className={classNames('co-status-card__health-item', className)}>
+      <div className={classNames('co-status-card__health-item', className)} data-test={dataTest}>
```
cant we just generate the data-test value to be something like
```suggestion
      <div className={classNames('co-status-card__health-item', className)} data-test={`${title}-health-item`}>
```

we wouldnt need to pass the `dataTest` prop

rexagod replied:
Fixed!

rawagner commented:
/lgtm
